ext/odbc/php_odbc.c

```diff
@@ -49,23 +51,205 @@
 
 #include "odbc_arginfo.h"
 
-/*
- * not defined elsewhere
- */
+#define CHECK_ODBC_CONNECTION(conn) \
+	if (conn == NULL) { \
+		zend_throw_error(NULL, "ODBC connection has already been closed"); \
+		RETURN_THROWS(); \
+	}
 
-#ifndef TRUE
-#define TRUE 1
-#define FALSE 0
-#endif
+#define CHECK_ODBC_RESULT(result) \
+	if (result->conn_ptr == NULL) { \
+		zend_throw_error(NULL, "ODBC result has already been closed"); \
+		RETURN_THROWS(); \
+	}
 
 void odbc_do_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent);
+static void safe_odbc_disconnect(void *handle);
+static void close_results_with_connection(odbc_connection *conn);
+static inline odbc_result *odbc_result_from_obj(zend_object *obj);
+
+static int le_pconn;
+
+static zend_class_entry *odbc_connection_ce, *odbc_result_ce;
+static zend_object_handlers odbc_connection_object_handlers, odbc_result_object_handlers;
+
+#define Z_ODBC_LINK_P(zv) odbc_link_from_obj(Z_OBJ_P(zv))
+#define Z_ODBC_CONNECTION_P(zv) Z_ODBC_LINK_P(zv)->connection
+#define Z_ODBC_RESULT_P(zv) odbc_result_from_obj(Z_OBJ_P(zv))
+
+static void odbc_insert_new_result(odbc_connection *connection, zval *result) {
+	ZEND_ASSERT(Z_TYPE_P(result) == IS_OBJECT);
+#if ZEND_DEBUG
+	ZEND_ASSERT(instanceof_function(Z_OBJCE_P(result), odbc_result_ce));
+#endif
+
+	odbc_result *res = Z_ODBC_RESULT_P(result);
 
-static int le_result, le_conn, le_pconn;
+	res->index = connection->results.nNextFreeElement;
+	zend_hash_index_add_new(&connection->results, res->index, result);
+	Z_ADDREF_P(result);
+}
+
+static inline odbc_link *odbc_link_from_obj(zend_object *obj) {
+	return (odbc_link *)((char *)(obj) - XtOffsetOf(odbc_link, std));
+}
+
+static int _close_pconn_with_res(zval *zv, void *p)
+{
+	zend_resource *le = Z_RES_P(zv);
+	if (le->type != le_pconn) {
+		return ZEND_HASH_APPLY_KEEP;
+	}
+
+	odbc_connection *list_conn = ((odbc_connection *)(le->ptr));
```

Comment:
Yes, initially I had a similar code which you suggested, but I wanted to debug it, that's why I created the local variables so that I can inspect them. Now that finally I don't have to do debugging, I'm fine to revert these changes